metadb/server.go

```diff
@@ -0,0 +1,233 @@
+	"fmt"
+	"net"
+	"path/filepath"
+	"time"
+
+	"github.com/pastelnetwork/gonode/common/errors"
+	"github.com/pastelnetwork/gonode/common/log"
+	"github.com/pastelnetwork/gonode/metadb/rqlite/cluster"
+	httpd "github.com/pastelnetwork/gonode/metadb/rqlite/http"
+	"github.com/pastelnetwork/gonode/metadb/rqlite/store"
+	"github.com/pastelnetwork/gonode/metadb/rqlite/tcp"
+)
+
+const (
+	defaultJoinAttempts           = 5
+	defaultJoinInterval           = 5 * time.Second
+	defaultRaftHeartbeatTimeout   = time.Second
+	defaultRaftElectionTimeout    = time.Second
+	defaultRaftApplyTimeout       = 10 * time.Second
+	defaultRaftOpenTimeout        = 60 * time.Second
+	defaultRaftWaitForLeader      = true
+	defaultRaftSnapThreshold      = 8192
+	defaultRaftSnapInterval       = 30 * time.Second
+	defaultRaftLeaderLeaseTimeout = 0 * time.Second
+	defaultCompressionSize        = 150
+	defaultCompressionBatch       = 5
+)
+
+// determine the join addresses
+func (s *service) determineJoinAddresses(_ context.Context) ([]string, error) {
+	var addrs []string
+
+	// <TODO> get the list of supernodes from Pastel RPC API, and try to connect automatically.
+
+	return addrs, nil
+}
+
+// wait until the store is in full consensus
+func (s *service) waitForConsensus(ctx context.Context, dbStore *store.Store) error {
+	if _, err := dbStore.WaitForLeader(defaultRaftOpenTimeout); err != nil {
+		if defaultRaftWaitForLeader {
+			return errors.Errorf("leader did not appear within timeout: %w", err)
+		}
+		log.WithContext(ctx).Infof("ignoring error while waiting for leader")
+	}
+	if err := dbStore.WaitForApplied(defaultRaftOpenTimeout); err != nil {
+		return errors.Errorf("store log not applied within timeout: %w", err)
+	}
+
+	return nil
+}
+
+// start the http server
+func (s *service) startHTTPServer(ctx context.Context, dbStore *store.Store, cs *cluster.Service) error {
+	httpAddr := fmt.Sprintf("%s:%d", s.config.ListenAddress, s.config.HTTPPort)
+	// create http server
+	server := httpd.New(ctx, httpAddr, dbStore, cs, nil)
+
+	// start the http server
+	return server.Start()
+}
+
+// start a mux for rqlite node
+func (s *service) startNodeMux(ctx context.Context, ln net.Listener) (*tcp.Mux, error) {
+	mux, err := tcp.NewMux(ctx, ln, nil)
+	if err != nil {
+		return nil, errors.Errorf("create node-to-node mux: %w", err)
+	}
+
+	go mux.Serve()
+
+	return mux, nil
+}
+
+// start the cluster server
+func (s *service) startClusterService(ctx context.Context, tn cluster.Transport) (*cluster.Service, error) {
+	c := cluster.New(ctx, tn)
+
+	httpAddr := fmt.Sprintf("%s:%d", s.config.ListenAddress, s.config.HTTPPort)
+	// set the api address
+	c.SetAPIAddr(httpAddr)
+
+	// open the cluster service
+	if err := c.Open(); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// create and open the store of rqlite cluster
+func (s *service) initStore(ctx context.Context, raftTn *tcp.Layer) (*store.Store, error) {
+	raftDir := filepath.Join(s.workDir, s.config.DataDir)
```

Comment:
I meant different. 
By default it should be like this

take an example from 
https://github.com/pastelnetwork/gonode/blob/add54140e3bcb36806bad451d14c37a115cf920c/p2p/config.go#L27-L32
https://github.com/pastelnetwork/gonode/blob/add54140e3bcb36806bad451d14c37a115cf920c/supernode/cmd/app.go#L133

but an user can specify another dir, anywhere in file system. 
for example, to place the database to a raid disk, or whatever.